### PR TITLE
[LTR backport] Fix qgsinterpolator and qgstininterpolator with NULL value (#43302)

### LINKED ISSUE #42689: Error in TIN interpolation with null values
**Describe the bug**
When doing a tin interpolation on a layer that has fields with data and null values, the null values are considered as having a value of 0 instead of not being considered for the interpolation.
I have made several tests with versions 3.18.1, 3.16.5 and 3.10.12 that I have installed and it has only worked correctly in version 3.10.12.

TIN interpolation

**How to Reproduce**

TIN interpolation
Sample file included in ZIP. Tested with the sample file and other files.

**QGIS and OS versions**

3.18.1, ERROR
 3.16.5  ERROR
3.10.12 OK

**Additional context**

### LINKED ISSUE #43286: Error in IDW interpolation with null values
**Describe the bug**
Time ago I opened the BUG: [https://github.com/qgis/QGIS/issues/42689](url) related with TIN interpolation.
Today, I realised that IDW interpolation has the same bug. The problem is null values are considered as 0 instead null.
In the screenshot the centerpoint has no value. All the others has a value = 5.
3.18.3, ERROR
3.16.5 ERROR
3.10.12 OK

[IDW-BUG.zip](https://github.com/qgis/QGIS/files/6510321/IDW-BUG.zip)

image

### PR DESCRIPTION
**Pull request by rouault**
Fixes #43286, fixes #42689

Backport of PR #43302

### CODE REVIEW

File: tests/src/analysis/testqgsinterpolator.cpp [hunk 1]
```diff
@@ -280,6 +286,160 @@ void TestQgsInterpolator::dualEdge()
   QVERIFY( QgsMesh::compareFaces( mesh.face( 3 ), QgsMeshFace( {2, 4, 1} ) ) );
 }
 
+void TestQgsInterpolator::TIN_IDW_Interpolator_with_Z()
+{
+  std::unique_ptr<QgsVectorLayer>mLayerPoint = std::make_unique<QgsVectorLayer>( QStringLiteral( "PointZ" ),
```
Comment (vcloarec) [inline tests/src/analysis/testqgsinterpolator.cpp:291]:
```suggestion
  std::unique_ptr<QgsVectorLayer>mLayerPoint = qgis::make_unique<QgsVectorLayer>( QStringLiteral( "PointZ" ),
```

**Comment by vcloarec** [comment]
thanks for the manual backport!

File: tests/src/analysis/testqgsinterpolator.cpp [hunk 2]
```diff
@@ -280,6 +286,160 @@ void TestQgsInterpolator::dualEdge()
...
+  QVERIFY( qgsDoubleNear( resultTIN, 3.0, 0.00000001 ) );
+  QVERIFY( qgsDoubleNear( resutlIDW, 2.94444444444, 0.00000001 ) );
+
+  //! Outside the hull, TIN interpolator has to return an error
+  QVERIFY( tin.interpolatePoint( 2.5, 1.5, resultTIN, nullptr ) != 0 );
+}
+
+void TestQgsInterpolator::TIN_IDW_Interpolator_with_attribute()
+{
+  std::unique_ptr<QgsVectorLayer>mLayerPoint = std::make_unique<QgsVectorLayer>( QStringLiteral( "Point?field=ZValue:real" ),
```
Comment (vcloarec) [inline tests/src/analysis/testqgsinterpolator.cpp:362]:
```suggestion
  std::unique_ptr<QgsVectorLayer>mLayerPoint = qgis::make_unique<QgsVectorLayer>( QStringLiteral( "Point?field=ZValue:real" ),
```

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Unrelated test failure

**Pull request merged**